hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/TransactionMetadata.java

```diff
@@ -15,61 +15,97 @@
+ *     to payer key
+ * @param readKeys the keys that were read during pre-handle
  */
-public interface TransactionMetadata {
-    /**
-     * Checks the failure by validating the status is not {@link ResponseCodeEnum OK}
-     *
-     * @return returns true if status is not OK
-     */
-    default boolean failed() {
-        return !status().equals(ResponseCodeEnum.OK);
+public record TransactionMetadata(
+        @Nullable TransactionBody txnBody,
+        @Nullable AccountID payer,
+        @NonNull ResponseCodeEnum status,
+        @Nullable HederaKey payerKey,
+        @NonNull List<HederaKey> requiredNonPayerKeys,
+        @Nullable Object handlerMetadata,
+        @NonNull List<ReadKeys> readKeys) {
+
+    public TransactionMetadata {
+        requireNonNull(status);
+        requireNonNull(requiredNonPayerKeys);
+        requireNonNull(readKeys);
     }
 
-    /**
-     * Returns the status of the transaction.
-     *
-     * @return the status of the transaction.
-     */
-    ResponseCodeEnum status();
+    public TransactionMetadata(
+            @NonNull final PrehandleHandlerContext context,
+            @NonNull final List<ReadKeys> readKeys) {
+        this(
+                requireNonNull(context).getTxn(),
+                context.getPayer(),
+                context.getStatus(),
+                context.getPayerKey(),
+                context.getRequiredNonPayerKeys(),
+                context.getHandlerMetadata(),
+                readKeys);
+    }
 
-    /**
-     * Transaction that is being pre-handled
-     *
-     * @return transaction that is being pre-handled
-     */
-    TransactionBody txnBody();
+    public TransactionMetadata(
+            @Nullable final TransactionBody txBody,
+            @Nullable final AccountID payerID,
+            @NonNull final ResponseCodeEnum status) {
+        this(txBody, payerID, status, null, List.of(), null, List.of());
+    }
 
     /**
-     * All the keys required for validation signing requirements in pre-handle. This list includes
-     * payer key as well.
+     * Checks the failure by validating the status is not {@link ResponseCodeEnum OK}
      *
-     * @return keys needed for validating signing requirements
+     * @return returns true if status is not OK
      */
-    List<HederaKey> requiredNonPayerKeys();
+    public boolean failed() {
+        return status != ResponseCodeEnum.OK;
+    }
 
     /**
-     * Payer for the transaction
+     * An entry of read keys for a single {@link com.hedera.node.app.spi.state.ReadableKVState}
      *
-     * @return payer for the transaction
-     */
-    AccountID payer();
-
-    /**
-     * Transaction payer's key
+     * <p>Each entry in the list consists of the {@code statesKey} (which identifies the {@link
+     * com.hedera.node.app.spi.state.ReadableStates}), the {@code stateKey} (which identifies the
+     * {@link com.hedera.node.app.spi.state.ReadableKVState}, and the {@link Set} of keys, that were
```

Comment:
Need to add more javadoc to avoid confusion on `keys`, which mean account numbers

Reply:
Right, let's put it on the list of topics to discuss tomorrow. Once we have agreed on the wording, documenting it will be the easy part. 😄